### PR TITLE
feat(ui-shell): design updates to side nav

### LINKED ISSUE #2402: Development of standalone Left Panel
Design issue: https://github.com/carbon-design-system/carbon-components/issues/2335

### LINKED ISSUE #2703: Showcase different Left Panel patterns in storybook
Add in variations of all the different UI Shell Global Header patterns that are supported

### PR DESCRIPTION
**Pull request by tw15egan**
Closes #2402 
Closes #2703 
Refs #2335 

This PR contains design/spec updates to the fixed SideNav component. UX patterns/interactions will be added in a separate PR (#2801)

image

### CODE REVIEW

**Comment by tw15egan** [comment]
cc @aledavila @jnm2377

**Comment by tw15egan** [comment]
Most of the design changes have been implemented, currently working on how to best bake in the toggling ability of the side nav

@asudoh perhaps you have some suggestions? 

I created a `Fixed SideNav w/ Header` story, where the `SideNav` resides inside of `Header`. Wondering what's the best way to toggle the expanded state when they both exist.

**Comment by asudoh** [comment]
Seems that the latest version of side nav acts like our old hamburger left nav rather than our old ILN. Is it correct? If so, we can add a new prop to `<SideNav>` that dictates the expanded state.

`<Header>` could manage the corresponding React state. Doing so creates a new requirement that lets application control such React state. The current consensus in the team for that is introducing a pair of properties, say `isSideNavExpanded`/`onSideNavToggle`, and do something like:

```javascript
const { isSideNavExpanded, onSideNavToggle } = this.props;
const [isSideNavExpandedState, setIsSideNavExpandedState] = useState(isSideNavExpanded);
const handleHeaderMenuButtonClick = useCallback(() => {
  const newIsSideNavExpandedState = !isSideNavExpandedState;
  if (onSideNavToggle) {
    onSideNavToggle({ isSideNavExpanded: newIsSideNavExpandedState });
  } else {
    setIsSideNavExpandedState(newIsSideNavExpandedState);
  }
}, [isSideNavExpandedState, setIsSideNavExpandedState, onSideNavToggle]);
...
<HeaderMenuButton
  aria-label="Open menu"
  onClick={handleHeaderMenuButtonClick}
/>
...
<SideNav isExpanded={onSideNavToggle ? isSideNavExpanded : isSideNavExpandedState}>
```

**Comment by tw15egan** [comment]
For the sake of progress, I think I'm going to split the UX / interaction updates into a separate issue / PR. This one will simply handle the fixed side nav design updates

**Comment by tw15egan** [comment]
Issue: https://github.com/carbon-design-system/carbon/issues/2801

File: packages/components/src/components/ui-shell/_side-nav.scss
Comment (emyarod) [inline packages/components/src/components/ui-shell/_side-nav.scss]: `&` here is redundant right?

**Review by emyarod — COMMENTED** [review]
is this PR for react only?

vanilla example:

image

**Comment by tw15egan** [comment]
Changes were only in React, but I can make sure we have an example of the Vanilla Fixed Side Nav. Side Rail will be next iteration, so it may be a bit broken until then.

**Review by jnm2377 — APPROVED** [review]
This looks good to me. We just need to add the hamburger menu interaction with left nav in a different PR right?

**Comment by tw15egan** [comment]
@jnm2377 That's correct. Once this gets merged we can start working on that. Going to update the examples for the vanilla version now, and then we should be good to go

**Comment by tw15egan** [comment]
@emyarod updated fixed side nav example

**Comment by emyarod** [comment]
nice @tw15egan vanilla and react fixed side navs look identical now. should we include some example menu items with icons similar to the spec screenshot?

**Comment by tw15egan** [comment]
Yeah, can we track that in a separate issue? Don’t want this to hold up @jnm2377

**Comment by designertyler** [comment]
The chevrons are 20 x 20px and shrink to 16 x 20px when a category title is truncated.
The chevrons should be 16 x 16px for all states.
image

A selected submenu's text (the "L0 menu item") weight should be: 600 / Semi-Bold The design specs have it mocked up correctly, but incorrectly say it's 400 / regular. Sorry about that.

image

I'm getting a weird behavior when I click on a submenu item. The active state should only be on the selected menu item and not the whole category.

May-28-2019 14-04-24


When a submenu is selected the parent should get the 600 / Semi-Bold too. This helps the user know where they are even if they collapse the menu.
image

**Comment by tw15egan** [comment]
@designertyler comments are addressed

**Review by dakahn — APPROVED** [review]

**Comment by designertyler** [comment]
Changes look good.

I just noticed the side nav submenu text moves 2px to the right when it's truncated (top example). It only seems to happen when there's a caret for the submenu. The normal side nav text stays aligned (bottom example). Maybe there's something happening with the chevron icon?  

image


**Edit: one more**

When a submenu item is selected and then the menu is collapsed the parent should get the selected state. Similar to our website's behavior. It helps the user know where they are even if the menu is collapsed.

image

Like this:
May-29-2019 10-32-40

**Comment by tw15egan** [comment]
@designertyler updated!

**Review by designertyler — APPROVED** [review]
Looks good!

**Comment by asudoh** [comment]
🎉 @tw15egan Would you be able to solve the merge conflict? Thanks!

**Comment by aledavila** [comment]
@tw15egan thanks. For the merge conflict. There's no longer a hover on an active action.

**Comment by tw15egan** [comment]
@jnm2377 and @aledavila just rebased both your PR's on to this one. Since it was a lot of commits I just want to make sure I didn't accidentally remove anything 👀

File: packages/react/src/components/UIShell/UIShell-story.js
Comment (jnm2377) [inline packages/react/src/components/UIShell/UIShell-story.js]: This is not a big deal, but I had updated this in my PR just so that the example looks better. The lorem without spacing and type styles is jarring. But also since I used the grid as part of the content, I was trying to show how the content can use an offset class when the sidenav is persistent, and then when it collapses the content takes up the full width.

File: packages/react/src/components/UIShell/UIShell-story.js
Comment (jnm2377) [inline packages/react/src/components/UIShell/UIShell-story.js]: I added a prop to `SideNav` called `isChildOfHeader`, which will accept a bool. This adds specific styles for the header/sidenav ux. I set it as true by default (we can change that if we want), but the example for fixed `SideNav` is lower by 48px since it doesn't have `isChildOfHeader={false}` so it's adding the header-ux styles.

File: packages/react/src/components/UIShell/UIShell-story.js
Comment (tw15egan) [reply packages/react/src/components/UIShell/UIShell-story.js]: Whoops! Definitely didn't mean to remove that. Good catch 👍

File: packages/react/src/components/UIShell/UIShell-story.js
Comment (tw15egan) [reply packages/react/src/components/UIShell/UIShell-story.js]: Updated 👍

**Review by aledavila — APPROVED** [review]
lgtm - doesn't affect my panels

**Pull request merged**
